### PR TITLE
[v0.10.x] Softmax optimization & bertpass refactor

### PR DESCRIPTION
**Pull request by bgawrych**
This PR adds graph pass to optimize CPU's softmax on BERT.
Currently for BERT-large length tensor is created by following operations:  expand_dims -> brodcast_axis -> Reshape
and there is x24 such tensor creation. This pass replace softmax (with length) with regular softmax (but with masked input) - mask is created only once and then is passed to elemwise_sum to mask input. Applying pass in the scripts is optional

Original:
image

Masked softmax:
image

Thoughput in samples/s:

batches | batch_size | fp32 | quantized | quantized +   mha_interleave | quantized +   mask_softmax | quantized +   mask_softmax + mha_interleave
-- | -- | -- | -- | -- | -- | --
1000 | 24 | 19,74 | 25,08 | 25,26 | 34,36 | 34,80
500 | 1 | 14,06 | 16,97 | 18,02 | 20,24 | 21,81

Accuracy:

batch_size | fp32 | quantized | quantized +   mha_interleave | quantized +   mask_softmax | quantized +   mask_softmax + mha_interleave
-- | -- | -- | -- | -- | --
EM | 80,99 | 80,91 | 80,44 | 80,73 | 80,44
F1 | 88,60 | 88,33 | 88,06 | 88,29 | 88,06

There is also bug fix in interrleaved mha pass
Accuracy without mha_interleave bug fix: {'exact_match': 79.62157048249763, 'f1': 87.75497143592598}

### CODE REVIEW

**Comment by bartekkuncer** [comment]
LGTM

File: scripts/bert/bertpass.cc [hunk 1]
```diff
@@ -460,14 +496,23 @@ MXReturnValue custom_pass(mxnet::ext::Graph *g,
...
+      counter = 0;
+      for (int e = 1; e < num_heads * 3; e += 3) {
+          for (int h = e * head_dimension; h < e * head_dimension + head_dimension; h++) {
+            qkv_bias_data[h] = key_bias_data[counter++];
+          }
       }
-      for (int e=0; e < shape0; ++e) {
-        qkv_bias_data[2 * shape0 + e] = value_bias_data[e];
+      counter = 0;
+      for (int e = 2; e < num_heads * 3; e += 3) {
```
Comment (mozga-intel) [inline scripts/bert/bertpass.cc:512]: Could you please tell me whether we can avoid proliferation of those loops? It looks like that these loops are "identical", only a few parameters are different.

File: scripts/bert/bertpass.cc [hunk 2]
```diff
@@ -476,14 +521,235 @@ MXReturnValue custom_pass(mxnet::ext::Graph *g,
...
+  if (reshape_node->op == "Reshape") {
+    std::string required_shape = reshape_node->attrs["shape"];
+    std::string is_reverse = reshape_node->attrs["reverse"];
+    if (required_shape != "(-1, 0)" || is_reverse != "True") {
+      return false;
+    }
+  } else {
+    return false;
+  }
+  auto bcast_axis_node = reshape_node->inputs[0].node;
```
Comment (mozga-intel) [inline scripts/bert/bertpass.cc:557]: Please add a white line [between 556 & 557].

File: scripts/bert/bertpass.cc [hunk 3]
```diff
@@ -460,14 +496,23 @@ MXReturnValue custom_pass(mxnet::ext::Graph *g,
       }
 #endif
       // concatenate bias terms
-      for (int e =0; e < shape0; ++e) {
-        qkv_bias_data[e] = query_bias_data[e];
+      int counter = 0;
+      for (int e = 0; e < num_heads * 3; e += 3) {
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply scripts/bert/bertpass.cc]: 3 comes from query, key, value

File: scripts/bert/bertpass.cc [hunk 4]
```diff
@@ -460,14 +496,23 @@ MXReturnValue custom_pass(mxnet::ext::Graph *g,
...
+          for (int h = e * head_dimension; h < e * head_dimension + head_dimension; h++) {
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply scripts/bert/bertpass.cc]: Done

File: scripts/bert/bertpass.cc (hunk 1 above)
Comment (bgawrych) [reply scripts/bert/bertpass.cc]: Done, please check

File: scripts/bert/index.rst [hunk 5]
```diff
@@ -348,7 +357,7 @@ Once the model is exported, you can import the model by setting --only_infer, an
 The batch size can be specified via --test_batch_size option, and accuracy can be checked setting --check_accuracy.
 
 When using GPU and data type FP16 (--dtype float16), we recommend to use MXNET_FC_TRUE_FP16=1 for boosting performance.
-Moreover, you can use a custom graph pass for BERT, via --custom_pass [custom_pass_file], to improve the performance on GPU. To generate the pass you can run setup.py within the BERT scripts directory. These GPU optimizations require MXNet version 1.7 or higher.
+Moreover, you can use a custom graph pass for BERT, via --custom_pass_lib [custom_pass_library] and --custom_passes [space_seperated_names_of_passes_to_apply], to improve the performance on GPU. To generate the pass you can run setup.py within the BERT scripts directory. These GPU optimizations require MXNet version 1.7 or higher.
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply scripts/bert/index.rst]: Done in edited fragments

File: scripts/bert/finetune_classifier.py [hunk 6]
```diff
@@ -698,8 +743,9 @@ def evaluate(loader_dev, metric, segment):
                         num_calib_batches,
                         quantized_dtype,
                         calib_mode)
-        except AttributeError:
+        except AttributeError as e:
+            warnings.warn(e)
             nlp.utils.version.check_version('1.7.0', warning_only=True, library=mx)
-            warnings.warn('INT8 Quantization for BERT need mxnet-mkl >= 1.6.0b20200115')
+            warnings.warn('INT8 Quantization for BERT need mxnet-mkl >= 1.6.0b20200115 or mxnet >= 1.7')
```
Comment (szha) [inline scripts/bert/finetune_classifier.py:749]: since the version check checks for 1.7, there's no need to keep the 1.6 nightly version in the warning.

**Comment by bgawrych** [comment]
@szha Can you help with CI? I'm not sure why it's failing and don't know how to rerun it.
website-build seems to fail on notebook I haven't edited

**Comment by barry-jin** [comment]
Hi @bgawrych, Could you try to merge with v0.10.x, we ported new CI settings from v0.x to v0.10.x. Thanks!

**Comment by bgawrych** [comment]
@barry-jin, @szha still issue with notebook, little strange as bert.md file was not changed for 1 year - should I fix it or it's CI issue?

**Comment by leezu** [comment]
@barry-jin I see the following error in the log, pointing out that there's a CI issue:

```
[2021-05-10T05:59:14.916Z] umount: /dev/shm: must be superuser to unmount.
[2021-05-10T05:59:14.917Z] mount: /dev/shm: permission denied.
[2021-05-10T05:59:14.917Z] ./gluon_nlp_job.sh: line 33: sudo: command not found
```

**Comment by barry-jin** [comment]
> @barry-jin, @szha still issue with notebook, little strange as bert.md file was not changed for 1 year - should I fix it or it's CI issue?

Hi @bgawrych, we have ported the changes in bert.md from v0.x branch, you could try to merge with current v0.10.x.

**Comment by barry-jin** [comment]
> @barry-jin I see the following error in the log, pointing out that there's a CI issue:
> 
> ```
> [2021-05-10T05:59:14.916Z] umount: /dev/shm: must be superuser to unmount.
> [2021-05-10T05:59:14.917Z] mount: /dev/shm: permission denied.
> [2021-05-10T05:59:14.917Z] ./gluon_nlp_job.sh: line 33: sudo: command not found
> ```

Thanks, I will fix this issue.

**Comment by bgawrych** [comment]
@barry-jin 
```
[2021-05-19T09:19:56.267Z] Exception occurred:
[2021-05-19T09:19:56.267Z]   File "/workspace/gluon-nlp/docs/conf.py", line 237, in setup
[2021-05-19T09:19:56.267Z]     app.add_javascript('google_analytics.js')
[2021-05-19T09:19:56.267Z] AttributeError: 'Sphinx' object has no attribute 'add_javascript'
```

**Comment by barry-jin** [comment]
> @barry-jin
> 
> ```
> [2021-05-19T09:19:56.267Z] Exception occurred:
> [2021-05-19T09:19:56.267Z]   File "/workspace/gluon-nlp/docs/conf.py", line 237, in setup
> [2021-05-19T09:19:56.267Z]     app.add_javascript('google_analytics.js')
> [2021-05-19T09:19:56.267Z] AttributeError: 'Sphinx' object has no attribute 'add_javascript'
> ```

Will be fixed in https://github.com/dmlc/gluon-nlp/pull/1575

**Review by szha — APPROVED** [review]

**Comment by szha** [comment]
Merged. Thanks @bgawrych!

**Comment by ptrendx** [comment]
I'm suprised that the elimination of the 24x mask tensor creation gave you any speedup (as opposed to using masked softmax, which should) - MXNet already has common expression elimination pass (I wrote it: https://github.com/apache/incubator-mxnet/pull/15657). Does that not work for you?

**Comment by bgawrych** [comment]
@ptrendx I didn't know about this feature, but I wrote this small graph pass to test it:

```
#if MX_LIBRARY_VERSION <= 7
MXReturnValue TEST(const std::string& in_graph, const std::string** out_graph,
                          const std::unordered_map<std::string, std::string>& options,
                          const std::unordered_map<std::string, MXTensor>& args,
                          const std::unordered_map<std::string, MXTensor>& aux,
                          const PassResource& res) {
  Graph *g = Graph::fromString(in_graph);
#else
MXReturnValue TEST(mxnet::ext::Graph *g,
                          const std::unordered_map<std::string, std::string>& options) {
#endif

  Node* commonnode;
#if MX_LIBRARY_VERSION <= 7
  for(Node* n : g->nodes) {
#else
  for(int i=0; i < g->size(); i++) {
    Node* n = g->getNode(i);
#endif
    if (n->op.compare("softmax") == 0) {
      commonnode = n->inputs[1].node;
      break;
    }
  }

#if MX_LIBRARY_VERSION <= 7
  for(Node* n : g->nodes) {
#else
  for(int i=0; i < g->size(); i++) {
    Node* n = g->getNode(i);
#endif
    if (n->op.compare("softmax") == 0) {
      n->inputs[1].node = commonnode;
    }
  }

#if MX_LIBRARY_VERSION <= 7
  // convert back to JSON string from Graph/Node
  *out_graph = new std::string(g->toString());
#endif
  return MX_SUCCESS;
}
```

Overhead from these operators are negligible, but seems like it don't work in this case:
with graph pass:
image



w/o
image

**Pull request merged**